### DISCUSSION TITLE
Force column order during table creation

### DISCUSSION THREAD

**Discussion by CyMathew** [simolus3/drift#2896] [Q&A] [answered]
Is there a way to force a certain ordering of columns when tables are created?
I'm using mixins to add columns to multiple tables for metadata/code reuse purposes.

Suppose I have a table like this:
```dart
class ChatEntries extends Table with MetaColumns {
  TextColumn get chatId => text()();
  TextColumn get chatText => text()();
  IntColumn get readStatus => integer().withDefault(const Constant(0))();

  @override
  Set<Column<Object>>? get primaryKey => {chatId}
}
```

With a simple mixin like this:

```dart
mixin MetaColumns on Table {
  TextColumn get userId => text()();
  DateTimeColumn get created => dateTime()();
}
```

Would it be possible to inform drift code_gen to put the mixin columns at the end? As it is right now, it would put `user_id` and `created` columns at the beginning of the table structure.

Probably a silly question but wanted to make the developer experience when debugging table data later a bit better.

**Comment by simolus3** [comment]
There's no good way to do this at the moment. You could write a class extending `$ChatEntriesTable` and overriding `$columns` to re-order them. You can then pass that subclass to `Migrator.createTable()` (without really having to use it anywhere else). But that's not too practical since you'd have to do it for every class where you want to change the order of columns.

**Comment by North101** [comment, marked as answer]
Would this work?

```
mixin MetaColumns on Table {
  TextColumn get userId => text()();
  DateTimeColumn get created => dateTime()();
}

mixin ChatEntriesColumns on Table {
  TextColumn get chatId => text()();
  TextColumn get chatText => text()();
  IntColumn get readStatus => integer().withDefault(const Constant(0))();
}

class ChatEntries extends Table with MetaColumns, ChatEntriesColumns {
  @override
  Set<Column<Object>>? get primaryKey => {chatId}
}
```

**Reply by CyMathew** [reply]
This certainly works and seems so obvious. Although, some may hate this solution for being very Java-like in abusing the interface-ness of mixins (You can move the `primaryKey` override to the mixin so the actual Table class would be entirely empty). Arguably trades developer experience troubles on DB table column order with some not-so-straightforward code and the chance to make a mess of everything by mixing in the wrong mixins to the wrong Table class.

But it works so I approve :)

The final resulting table class would be:

```dart
class ChatEntries extends Table with ChatEntriesColumns, MetaColumns {
}

mixin ChatEntriesColumns on Table {
  TextColumn get chatId => text()();
  TextColumn get chatText => text()();
  IntColumn get readStatus => integer().withDefault(const Constant(0))();

  @override
  Set<Column<Object>>? get primaryKey => {chatId}
}

mixin MetaColumns on Table {
  TextColumn get userId => text()();
  DateTimeColumn get created => dateTime()();
}

```
